libgit/autogit_manager.go

```diff
@@ -119,6 +130,144 @@ func (am *AutogitManager) getNewConfigDefault(ctx context.Context) (
+// the mdserver's time.  It's a very crude way of normalizing the
+// local clock.
+func (am *AutogitManager) commonTime(ctx context.Context) time.Time {
+	offset, haveOffset := am.config.MDServer().OffsetFromServerTime()
+	if !haveOffset {
+		am.log.CDebugf(ctx, "No offset, cannot use common time; "+
+			"falling back to local time")
+		return am.config.Clock().Now()
+	}
+	return am.config.Clock().Now().Add(-offset)
+}
+
+func (am *AutogitManager) canWorkOnRepo(
+	ctx context.Context, dstFS *libfs.FS, repo string) (
+	canWork bool, err error) {
+	am.log.CDebugf(ctx, "Checking if we can work on %s", repo)
+	defer func() {
+		am.deferLog.CDebugf(ctx, "Work check completed: canWork=%t, %+v",
+			canWork, err)
+	}()
+
+	// Take the lock for the dst repo while checking the work time.
+	lockFile, err := dstFS.Create(autogitLockName(repo))
+	if err != nil {
+		return false, err
+	}
+	defer func() {
+		// Because we took the lock, this Close will sync/flush the
+		// whole journal.
+		closeErr := lockFile.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+	err = lockFile.Lock()
+	if err != nil {
+		return false, err
+	}
+
+	// See if someone else is already working on this repo by seeing
+	// if the timestamp (converted to "common" time) is within the
+	// expected time limit for a worker.
+	workingFileName := autogitWorkingName(repo)
+	fi, err := dstFS.Stat(workingFileName)
+	currCommonTime := am.commonTime(ctx)
+	if err != nil && !os.IsNotExist(err) {
+		return false, err
+	} else if os.IsNotExist(err) {
+		am.log.CDebugf(ctx, "Creating new working file for %s", repo)
+		f, err := dstFS.Create(workingFileName)
+		if err != nil {
+			return false, err
+		}
+		err = f.Close()
+		if err != nil {
+			return false, err
+		}
+	} else { // err == nil
+		modCommonTime := fi.ModTime()
+		if modCommonTime.Add(workTimeLimit).After(currCommonTime) {
+			am.log.CDebugf(ctx, "Other worker is still working on %s; "+
+				"modCommonTime=%s, currCommonTime=%s, workTimeLimit=%s",
+				repo, modCommonTime, currCommonTime, workTimeLimit)
+			// The other worker is still running within the time
+			// limit.
+			return false, nil
+		}
+		am.log.CDebugf(ctx, "Other work expired on %s; "+
+			"modCommonTime=%s, currCommonTime=%s, workTimeLimit=%s",
+			repo, modCommonTime, currCommonTime, workTimeLimit)
+	}
+
+	am.log.CDebugf(ctx, "Setting work common time to %s", currCommonTime)
+	err = dstFS.Chtimes(workingFileName, time.Time{}, currCommonTime)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (am *AutogitManager) workDoneOnRepo(
+	ctx context.Context, dstFS *libfs.FS, repo string, workErr error) (
+	err error) {
+	am.log.CDebugf(ctx, "Completing work on %s, workErr=%+v", repo, workErr)
+	defer func() {
+		am.deferLog.CDebugf(ctx, "Work done completed: %+v", err)
+	}()
+
+	// Take the lock for the dst repo while checking the work time.
+	lockFile, err := dstFS.Create(autogitLockName(repo))
+	if err != nil {
+		return err
```

Comment:
In case something else is _trying_ to lock at the same time, maybe this should retry with a backoff.

Reply:
The `Create` won't fail in that case, but the `Lock` below will block indefinitely.  @jzila does that cover the case you're thinking of?

Reply:
Yep, thanks.